### PR TITLE
Support array type in primary and unique key.

### LINKED ISSUE #1715: Support array type in primary key/unique key

### PR DESCRIPTION
**Pull request by Linkerist**
before this PR:
```
mysql> create table t1(id bigint not null, ids array<int>) primary key(id) distributed by hash(id);
ERROR 1064 (HY000): Invalid aggregate function 'REPLACE' for 'ids'
mysql> create table t2(id bigint not null, ids array<int>) unique key(id) distributed by hash(id);
ERROR 1064 (HY000): Invalid aggregate function 'REPLACE' for 'ids'
mysql>
```

after this PR:
```
mysql> create table t1(id bigint not null, ids array<int>) primary key(id) distributed by hash(id);
Query OK, 0 rows affected (0.11 sec)

mysql> desc t1;
+-------+------------+------+-------+---------+-------+
| Field | Type       | Null | Key   | Default | Extra |
+-------+------------+------+-------+---------+-------+
| id    | BIGINT     | No   | true  | NULL    |       |
| ids   | ARRAY<INT> | Yes  | false | NULL    |       |
+-------+------------+------+-------+---------+-------+
2 rows in set (0.00 sec)

mysql> create table t2(id bigint not null, ids array<int>) unique key(id) distributed by hash(id);
Query OK, 0 rows affected (0.13 sec)

mysql> desc t2;
+-------+------------+------+-------+---------+-------+
| Field | Type       | Null | Key   | Default | Extra |
+-------+------------+------+-------+---------+-------+
| id    | BIGINT     | No   | true  | NULL    |       |
| ids   | ARRAY<INT> | Yes  | false | NULL    |       |
+-------+------------+------+-------+---------+-------+
2 rows in set (0.01 sec)

mysql>
```
ATT, resolve #1715

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java [hunk 1]
```diff
@@ -174,7 +174,7 @@ public void analyze(boolean isOlap) throws AnalysisException {
                         String.format("Cannot specify aggregate function '%s' for key column '%s'", aggregateType,
                                 name));
             }
-            if (!aggregateType.checkCompatibility(type.getPrimitiveType())) {
+            if (!(aggregateType.checkCompatibility(type.getPrimitiveType()) || (aggregateType.isReplaceFamily() && type.isArrayType()))) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:177]: Add unit test in BE code paths to test the array.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 1 above)
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:177]: `|| (aggregateType.isReplaceFamily() && type.isArrayType())`
IMO, it's better to place this check inside `ggregateType.checkCompatibility`

**Comment by sduzh** [comment]
Could you please give more test results on the reads and writes on array columns?

**Review by decster — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 1 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java:177]: the checkCompatibility is a simply table driven method, it check primitive types, the business logic of isArrayType, isHllType, isBitmapType, et al. is processed in analyze method right here.

**Comment by sduzh** [comment]
@kangkaisen @Seaven PTAL

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java (hunk 1 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnDef.java]: changed.

**Review by chaoyli — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Pull request merged**